### DISCUSSION TITLE
Gradient of a tensor in Moose

### DISCUSSION THREAD

**Discussion by ritamcj** [idaholab/moose#17741] [Q&A Modules: General] [answered]
Hi everyone,
I am attempting to develop a dislocation density based crystal plasticity framework in the tensor mechanics module in Moose. To calculate the GND density (geometrically necessary dislocations) using Nye tensor (Nye (1953)), **I need to compute the gradient of the plastic deformation gradient tensor** (Fp) (Das et al. (2018)). Could anyone kindly guide me to the relevant resources in Moose? I have looked in the Phase field module in the solution of the Allen-Cahn equation (ACinterface.C), where the mobility gradient is calculated. But I am unable to figure out how the "RealGradient" method is being used to compute gradients. I am attempting to calculate the expressions (1)-(6) as attached.

Ritam Chatterjee

![IMG_20210427_210619__01](https://user-images.githubusercontent.com/81842276/116914555-83a2e880-ac68-11eb-9d0e-a1fa68a0916e.jpg)

**Comment by jiangwen84** [comment, marked as answer]
This is a great question. `Fp` is directly calculated at quadrature points and it is not usually convenient to get its derivatives. One way you can do is to project `Fp` from quad points to nodes and then interpolate it using gradient of shape functions. In MOOSE, you can possibly perform global/local L2 project or Least Square fitting to get the nodal `Fp` values. For local LSF projection, you can take a look at an example at https://mooseframework.inl.gov/help/finite_element_concepts/nodal_patch_recovery.html

**Reply by ritamcj** [reply]
Dear Wen,
Thank you so much for your prompt response. Are there any existing implementations of the nodal patch recovery technique in Moose? I was able to find some relevant literature (Pitts (2019)) and now have some idea of how to implement. If I understand correct, the gradient of the shape function is to be multiplied with Fp at the quadrature points and the nodal values will be calculated by taking a polynomial interpolation over all neighbouring quadrature points as shown in the thesis of Pitts (2019). I will attempt to implement and revert back to you in case there are more queries. 
Warm regards,
Ritam

**Reply by sapitts** [reply]
Hi Ritam,
When I calculated the gradient of the Fp tensor, I used the nodal patch recovery method that Wen linked to above to "push" the value of each Fp tensor component to the nodes, and then used the `coupledGradient` API to find the values of the gradient of each tensor component. More information on the `coupledGradient` API can be found here: https://mooseframework.inl.gov/docs/doxygen/moose/classCoupleable.html#a9c8ae645c4ed3f969de1345f2cc4fb27 and https://mooseframework.inl.gov/source/interfaces/Coupleable.html 

Certainly this approach wasn't computationally efficient, and hasn't been added to the official MOOSE repository. I believe that the MOOSE team is discussing different methods for these types of calculations. I'd recommend watching the Newsletter for updates, and perhaps Wen can provide additional details if he has more information, too.

Hope this helps some,
Stephanie

**Reply by ritamcj** [reply]
Hi Stephanie,
Thank you very much for your suggestions. I will work on this and keep an eye on the newsletter.
Warm regards,
Ritam

**Reply by ritamcj** [reply]
Hi Wen and Stephanie,
I tried various approaches to use CoupledGradient to obtain grad(Fp), but unsuccessfully. I am pasting the original code here:

                     const VariableGradient &
                     Coupleable::coupledGradient(const std::string & var_name, unsigned int comp) const
                     {
                     const auto * const var = getVarHelper<MooseVariableField<Real>>(var_name, comp);
                     if (!var)
                     {
                     _default_gradient.resize(_coupleable_max_qps);
                     return _default_gradient;
                     }
                     checkFuncType(var_name, VarType::Gradient, FuncAge::Curr);

                     if (!_coupleable_neighbor)
                     return (_c_is_implicit) ? var->gradSln() : var->gradSlnOld();
                     return (_c_is_implicit) ? var->gradSlnNeighbor() : var->gradSlnOldNeighbor();
                     }

Now, as my variable is Fp, which is a rank two tensor, can the argument "var_name" be defined as something other than a string? Fp is calculated in FiniteStrainUOBasedCP and I am reading it in my C file as "_fp(getMaterialPropertyByName<RankTwoTensor>("fp"))". I tried storing the components of Fp in an array and defining var_name as "DenseVector<double>" but this did not work.

Basically, how can I define Fp as a string in my input file such that it can be passed as an argument to coupledGradient? Or else, is it possible to pass a vector or tensor to coupledGradient without having to define in the input file?

Ritam

**Reply by dschwen** [reply]
You are mixing up variables and material properties. Those are not exchangable. You'll have to extract the tensor components into 9 variables (either use AuxVariables and nodal patch recovery - not fully implemented), or use regular variables (e.g. named Fp00 Fp01 Fp02 Fp10 ... Fp22) and use the `MaterialPropertyValue` kernel with a `RankTwoCartesianComponent` material (the kernel solves for a (nodal) variable field to match the property values at quadrature points). Then you can construct a RankThreeTensor from the gradients of your nine variables. We might want to consider making an action for this.

**Reply by ritamcj** [reply]
Hi Daniel,
Thank you for your valuable suggestions. I will attempt these two methods.
Warm Regards,
Ritam

**Reply by ritamcj** [reply]
Hi Daniel,
I attempted the auxvariable method and created auxvariables to access the components of Fp. Here, I am pasting patches of my code for only one component of Fp. The auxvariables were defined as:
```
[./gnd_evol_alpha]
    type = CP_DD_GND_generation
    variable_size = 28
    Fp00 = fp_00
  [../]
```

Then I read the strings in my C file to apply to the CoupledGradient API as:
```
for (p = 0; p < 3; ++p)
{
   _grad_Fp00[p] = &coupledGradient("Fp00", p);
}
```
where I have defined the strings as:
```
  params.addRequiredCoupledVar("Fp00", "Fp(0,0)");
```

In the class declaration, the gradient variable is defined as:  
```
    _grad_Fp00(3)
```
which is defined in the header file as:
```
 std::vector<const VariableGradient *> _grad_Fp00;
```
Through this method, I am able to read the Fp components **at only the first time step and not the evolved values afterwards**. To verify this, we performed a bending test (attached) and checked the GND density evolution across a three-element structure but there was no change. Could you kindly point out an error in implementation if possible? In my humble opinion, it is not possible to read the updated Fp from FiniteStrainUObasedCP at each time step by reading the auxvariables from the input file. Thank you so much for your time.

Warm regards,
Ritam

![Screenshot from 2021-05-27 23-03-00](https://user-images.githubusercontent.com/81842276/119871146-cd1ce580-bf3f-11eb-98d6-ea17e6f810a2.png)
![Uniaxial_tension_GND2_26052021](https://user-images.githubusercontent.com/81842276/119871230-de65f200-bf3f-11eb-93fb-6d29565a836c.png)
